bidirectionalmap/map.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bidirectionalmap
+
+import (
+	"k8s.io/utils/genericinterfaces"
+	"k8s.io/utils/set"
+)
+
+// BidirectionalMap is a bidirectional map.
```

Comment:
I would flesh this out more.  What is the semantic, what guarantees does it offer or not offer, what is it for (a real example)?  Also explaion which is the left-key and right-value vs. the left-value and right-key.

Naively I would expect BiDiMap[X, Y] to mean map[X]Y + Map[Y]X, but this seems to be the opposite

Reply:
I'm not sure what you mean by opposite, `map[X]Y + Map[Y]X` can be invalidated by have multiple keys in `map[x]y` map to the same y value. e.g.:

```
a = map[string]string
a["a"] = "1"
a["b"] = "1"
```
The reverse mapping must return a list of the keys of map `a` otherwise we have a lossy map.

Therefore, a bidirectional map must be two maps, each of which map to a list (or set) of values.